### DISCUSSION TITLE
Attiny104

### DISCUSSION THREAD

**Discussion by jafrado** [SpenceKonde/ATTinyCore#538] [Development Plans]
Any plans to support?

**Comment by SpenceKonde** [comment]
Nope, this core does not cover the AVRrc (reduced core) parts. My policy has always been ATtiny parts based on AVRe instruction set variant only (AVRxt parts covered by megaTinyCore) which are available in a version with at least 4k of flash. Some of the AVRrc parts are covered by https://github.com/technoblogy/attiny10core but not that one. 

I think parts with only 1k of flash are not a good match for the Arduino environment - you're basically going to be writing raw non-arduino-ish C since the overhead of the arduino API is too much to do much with in 1k, (In attiny10core, he just doesn't ship with much in the way of API functions - because the user experience with them is crap, you can't fit hardly anything on the chip.

**Reply by jafrado** [reply]
Thanks, that is great advice. My intention in using this was to add a proper UART to an Attiny85. You're right; any significant firmware (in Arduino environment) will consume all of the FLASH to the extent that it is not worth the effort. I guess these parts are simply for bare on the metal loops without an Arduino IDE.

**Reply by SpenceKonde** [reply]
Here's something you probably didn't know - you know what led me to start working on and (almost immediately) maintaining and making available an Arduino core? I was a total unqualified noob when I started maintaining ATTinyCore (well, I hadn['t merged in that one yet). But I had come to Arduino from Espruino, and I might not have known much, but I knew that software serial implementations sucked. But I *alsio* have an obsessive desire to use smaller parts rather than large ones, and to design everything myself (both of which are harder for classic megaAVR vs classic tinyAVR). The ATTiny84 was pretty great except that it didn't have any serial ports... but there was this semi-new part, the ATtiny841, that had TWO serial ports, was pin compatible with the 84 (SMD only, but I very quickly learned that drag soldering makes SOIC packages a breeze to solder, so that wasn't a problem), AND got you 2 extra PWM channels! It was perfect - but the Arduino IDE didnt support it. 

now, there is one thing that I hate more than learning baroque full featured IDEs like Atmel/Microchip studio and it's ilk.... and that's fucking around with toolchain setup configurations which is what I would have had to do to use neither Atmel Studio. I *still* have never managed to get any joy from the MP Studio official Microchip Or the MPLab SNAP, for that matter. There are a ton of things I think Arduino did a miserable job of, but making an IDE that had a lot of the things that an advanced person needed, without the clutter and feeping creaturism of "pro-fesh-on-al" IDEs that make them hard to learn in the same way that photo editing software and CAD tools are. 

Well, I found some other guy who had a half-working core, I fixed his digital I.O functions and added the third timer... and then i discovered the 1634 with 16k flash and again, the 2 serial ports... one thing led to another >.>

So desire to get proper serial ports on a tiny is what led ME to my position as lord of the tinies and from there to the Dx czar as well...

**Reply by jafrado** [reply]
That's cool and a great story. I am looking forward to getting your hardware. I built a few Attiny85 designs already and am using your library with great success. Yes, all of these IDE's really suck, best to have a Makefile and cross-compiler you can run from a shell prompt. Arduino is the only exception for me besides some digressions with Eclipse I've made over the years. I feel safest at a shell prompt with vi or emacs - I grew up on 8051/8086, PIC16/18/24, SPARC, MIPS (now PIC32) and have been doing lots of ARM development from M0 to A8/A53 last 10 years ... so I've come back full circle. 

Simplicity is the ultimate sophistication, when you get back to this stuff (almost bare on the metal), it's really nice to be able to drop down a core like this and turn code without all the tooling and meta-levels of software ...

**Reply by jafrado** [reply]
By the way, it's interesting to note that all of the PIC16/18 designs which have similar functionality are jailed by shitty auto-code generators, home/company-grown tools, and ultimately horrible IDE's. I would have thought with the ATMEL acquisition the less is more philosophy would have reigned some of this bullshit in, but hey, they teach Java as a programming language instead of C or assembler these days and a distinction is made between CS/EE ... (sigh).  It seems Arduino is one of the last bastions of hope for putting the simple IDE first and working code rules back in the driver seat ...

**Reply by SpenceKonde** [reply]
I don't get it either, I would think they'd want as many people as possible to know how to program them, so it was cheaper and easier for potential customers to get people with experience. Then again, in the closing years of the classic AVR era, did you ever look at the price of bare chips on AVR vs iother architectyures? Atmels parts were a HUGE ripoff in most cases! And yet people loved them. I think the free toolchain and exposure from Arduino is a part of that ;-) Now we're getting the best of both worlds though, as we get Microchip tier process tech and prices and the same instructionset we love, and same engineers designing the peripheral, plus they are starting from a clean slate and could sweep up a lot of the rubbish. 

These days? no, java died a 15+ years ago, it's legacy business logic and android phone app development and their saleries are well below industry average, nd have been for a while. A few years back, javascript was big, though it sounds like now it's moving towards the datascience and machine learning languages. And yet, for the Real Programmers, C has always been a reliable career. 

IMO EE and CS ARE completely different. The sort of stuff you learn in CS is not just programming. It's mathier.... I would be a horrible computer scientist, and most CS people, even really good ones, are a lot worse at EE/embedded. The skills are somewhat similar.... but not really.

**Comment by sleemanj** [comment]
FWIW, my fork of ATTinyCore supports 4/5/9 and 10https://github.com/sleemanj/optiboot/blob/master/dists/README.md#attiny4-attiny5-attiny9-attiny10The usefulness is, debatable, personally I think the Tiny13 (which my fork also supports) is a better bet in the arduino environment and I have certainly used that for a number of useful projects "written in arduino", but have only used the 4/5/9/10 when I was making the core support them, not since. ---- On Mon, 22 Mar 2021 05:50:21 +1300  ***@***.***  wrote ----
Any plans to support?

**Reply by jafrado** [reply]
Thanks, this is great! I think the issue for me will be code space, I can see how this would be really valuable for some specific use-cases where we would only need to twiddle a few wires.

**Comment by SpenceKonde** [comment]
I don't really see why there is even any interest within Arduino circles in such underwhelming processors with so little flash.... 
Pricing on the 104, SOIC14, is 72 cents (77 for the high temp ones), in qty 1 from digikey, 47/50 cents in quantities of 5k from Microchip. 
Now, pricing on the ATtiny1614, with 16x more flash, and all the whizzbang features of the 1-series (including a DAC!), and as mch RAM as an atmega328p (2048b)? It's gotta cost a few times that right? twice the price, minimum?... uhhh....  78 cents and 80 cents for normal and extended temperature for onsies from digikey, 52 and 53 cents from microchip in 5k qty. 6 cents! to get 16x the flash, 64 times the RAM, and to go from a meagre featureset to better peripherals than an ATmega328p - oh, and it already has full Arduino support; it is my usual default for small projects!
The ATTiny1604, which has only 1k of RAM and loses a few features vs the 1614 is actually cheaper than the 104! 46/48 cents @ 5k, 70 cents for qty1 from digikey. Also fully supported on Arduino. And the 204 (the 2k flash version of the 1604 -is 36 cents @ 5k. 

There is no reason to go out of your way to use a 104 (which seeing as there's no arduino support, you would need to do), and frankly, I don't think there's any reason for the 104 to _exist_. It was released around the same time as the 0/1-series too... yet has crapola classic style peripherals. If you gotta use a classic AVR instead of an AVRxt one, ffs, use one of the ones with some sort of redeeming quality. Maybe a 861 if yo lke differential ADC measurements (despite having only 10 bits of resolution, it still buries even the DA/DB parts for certain use cases, because you can go all the way to the power rails (actually slightly beyond) in differential mode while using a reference other than Vcc - most newer parts only work up to the reference voltage - and don't have programmable gain) , or an 841 for the combination of similarity to the 84 and one of the best pre-2016 peripheral loadouts, or even something like the 9/10 for the rice-grain size package... not something that combines tiny flash., negligible ram, and crap peripherals....

**Reply by jafrado** [reply]
Outstanding!!! I will make a part file for Eaglecad and use this device. Oh wow, you have a great dev board! Let me buy one ASAP. Thank you very much for all of your great work!!!

**Comment by SpenceKonde** [comment]
Much thanks!
